multiprofit/fit_psf.py

```diff
@@ -0,0 +1,271 @@
+import gauss2d.fit as g2f
+import numpy as np
+import time
+from typing import Dict, List, Type
+
+import lsst.pex.config as pexConfig
+
+from multiprofit.fit_catalog import CatalogExposureABC, CatalogFitterConfig, ColumnInfo
+from multiprofit.modeller import LinearGaussians, make_psfmodel_null, Modeller
+from multiprofit.psfmodel_utils import make_psf_source
+
+
+class CatalogExposurePsfABC(CatalogExposureABC):
+    def get_psf_image(self, source):
+        raise NotImplemented("Derived classes must implement")
+
+
+class CatalogPsfFitterConfig(CatalogFitterConfig):
+    """Configuration for MultiProFit PSF image fitter."""
+    sigmas = pexConfig.ListField[float](default=[1.5, 3], doc="Number of Gaussian components in PSF")
+
+    def rebuild_psfmodel(self, row: astropy.table.Row) -> g2f.PsfModel:
+        n_gaussians = len(self.sigmas)
+        idx_gauss_max = n_gaussians - 1
+        sigma_xs = [0.]*n_gaussians
+        sigma_ys = [0.]*n_gaussians
+        rhos = [0.]*n_gaussians
+        fracs = [1.]*n_gaussians
+
+        for idx in range(n_gaussians):
+            sigma_xs[idx] = row[f"{self.prefix_column}comp{idx + 1}_sigma_x"]
+            sigma_ys[idx] = row[f"{self.prefix_column}comp{idx + 1}_sigma_y"]
+            rhos[idx] = row[f"{self.prefix_column}comp{idx + 1}_rho"]
+            if idx != idx_gauss_max:
+                fracs[idx] = row[f"{self.prefix_column}comp{idx + 1}_fluxfrac"]
+        return g2f.PsfModel(
+            make_psf_source(sigma_xs=sigma_xs, sigma_ys=sigma_ys, rhos=rhos, fracs=fracs).components
+        )
+
+    def schema(self) -> List[ColumnInfo]:
+        columns = super().schema()
+        n_gaussians = len(self.sigmas)
+        idx_gauss_max = n_gaussians - 1
+
+        for idx_gauss in range(n_gaussians):
+            prefix_comp = f"comp{idx_gauss + 1}_"
+            columns_comp = [
+                ColumnInfo(key=f'{prefix_comp}sigma_x', dtype='f8', unit=u.pix),
+                ColumnInfo(key=f'{prefix_comp}sigma_y', dtype='f8', unit=u.pix),
+                ColumnInfo(key=f'{prefix_comp}rho', dtype='f8', unit=u.pix),
+            ]
+            if idx_gauss != idx_gauss_max:
+                columns_comp.append(ColumnInfo(key=f'{prefix_comp}fluxfrac', dtype='f8'))
+            columns.extend(columns_comp)
+
+        return columns
+
+    def setDefaults(self):
+        self.prefix_column = "mpf_psf_"
+
+
+class CatalogPsfFitter:
+    """ Fit a Gaussian mixture model to a pixelated PSF image.
+    """
+    @staticmethod
+    def _get_data(img_psf: np.array, gain: float = 1e5):
+        # TODO: Improve these arbitrary definitions
+        # Look at S/N of PSF stars?
+        background = np.std(img_psf[img_psf < 2*np.abs(np.min(img_psf))])
+        # Hacky fix; PSFs shouldn't have negative values but often do
+        min_psf = np.min(img_psf)
+        if not (background > -min_psf):
+            background = -1.1 * min_psf
+        img_sig_inv = np.sqrt(gain / (img_psf + background))
```

Comment:
1. PSFEx certainly can and often did return negative values for the PSF. I think Piff fixed this, but in its standard configuration makes the outskirts of the PSF noisier.
2. I'll add a docstring specifying that `img_psf` should be normalized. I did have a half-working method to smooth out the noisy negative parts [here](https://github.com/lsst-dm/multiprofit/blob/main/multiprofit/utils.py#L29), but it worked half by accident, simply zeroing out the faintest positive pixels which are typically in the outskirts of the PSF. It didn't actually smooth the PSF azimuthally in elliptical annuli as it probably should, so I left it out for now.

Reply:
This seems like a bug. Can @TallJimbo verify that this is expected? I thought for sure that the methods to create a PSF image checked for non-negativity, otherwise normalizing the PSF to sum to one doesn't make sense.

Reply:
I'm not surprised; the PSF models are going to have noise in them if they're no constraint preventing it, even if it's at a very low level.  Forcing them to be positive would have to be done with great care to avoid size biases in the PSF model, so I think having PSF consumers that can't handle slightly negative values deal with them seems appropriate.

Reply:
Ok. I have to think about whether or not this is something I have to worry about in scarlet now...

Reply:
I wouldn't think so, but if I were going to enforce positivity, I'd probably average out elliptical annuli starting from the edge of the box going in and zero them if they're still negative. The way I did it before was zeroing out enough of the lowest-value positive pixels to cancel out all the negative values, which indirectly had a similar effect.